Support k8s 1.16

douglas opened the pull request:
**What are you trying to accomplish with this PR?**

Support Kubernetes 1.16.

** Kubernetes 1.16 changes

- https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16
- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16

**How is this accomplished?**

-  Build/Test krane with k8s 1.16
- Update k8s templates to apps/v1
- [Update k8s-ci metrics-server to 0.3.6](https://github.com/Shopify/k8s-ci/pull/32)
- Upgrade metrics-server container to v0.3.6
- Add support for apps/v1 k8s API client
- Improve template generation detection for k8s 1.16
- Fallback to deprecated fields if count and lastTimestamp aren't present 

**What could go wrong?**

This is a big change in the sense that in order to support `kubernetes 1.16`, we needed to migrate to `v1`API for `DaemonSet`, `Deployment`,  `ReplicaSet`, and `StatefulSet`.

I think the problems that this PR could bring are related to the changes needed in user projects to support Kubernetes 1.16 and `v1` API.

dturn commented on test/helpers/fixture_set.rb:
```diff
@@ -28,7 +28,14 @@ def app_name
-      client = beta ? v1beta1_kubeclient : kubeclient
+      client = if beta
+        v1beta1_kubeclient
+      elsif type == 'deployment'
```
Should this include the other kinds that had to be upgraded: `StatefulSet, DaemonSet, ...`

dturn commented on test/integration/krane_deploy_test.rb:
```diff
@@ -1122,7 +1122,10 @@ def test_bad_container_on_stateful_sets_fails_with_rolling_update
-    result = deploy_fixtures("hello-cloud", subset: ["stateful_set.yml"])
+    result = deploy_fixtures("hello-cloud", subset: ["stateful_set.yml"]) do |fixtures|
+      stateful_set = fixtures['stateful_set.yml']['StatefulSet'].first
+      stateful_set['spec']['updateStrategy'] = { 'type' => Krane::StatefulSet::ONDELETE }
```
Is this because the default behavior changed?

douglas replied:
Yes, the default `updateStrategy`  for `StatefulSet API` for `v1beta1` is `OnDelete`, but in `v1` it is now `RollingUpdate`.

I added a comment to add more context in this section of the code.

For more information: 

- https://github.com/kubernetes/kubernetes/issues/56806#issuecomment-350197966
- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#statefulsetupdatestrategy-v1beta1-apps
- https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#statefulsetspec-v1-apps

douglas commented:
For the latest commit this is what it looks like:

```ruby
# kubernetes 1.16

[1] pry(Krane::KubernetesResource::Event)> FIELDS
=> [".involvedObject.kind", ".involvedObject.name", ".count", ".lastTimestamp", ".reason", ".message", ".eventTime"]
[2] pry(Krane::KubernetesResource::Event)> pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
=> ["Pod",
 "pvc",
 "<no value>",
 "<no value>",
 "FailedScheduling",
 "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
 "2019-11-06T03:03:10.351218Z"]
[3] pry(Krane::KubernetesResource::Event)>

# kubernetes 1.15.3

[1] pry(Krane::KubernetesResource::Event)> FIELDS
=> [".involvedObject.kind", ".involvedObject.name", ".count", ".lastTimestamp", ".reason", ".message", ".eventTime"]
[2] pry(Krane::KubernetesResource::Event)> pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
=> ["Pod",
 "pvc",
 "5",
 "2019-11-06T03:07:52Z",
 "FailedScheduling",
 "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
 "<no value>"]
```

That is why I'm using `.eventTime` as a fallback to `.lastTimestamp`.

dturn commented:
Looks like `lastTimestamp` was supported in Event v1 and becomes `deprecatedLastTimestamp ` for Events in v1beta1 https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/ (same for `count` and `deprecatedCount`) . I'm assuming that in 1.16 v1beta1 is now the preferred version from kubectl and that's why we're discovering this now. 

The long term fix is to parse the `series` field (which if empty means the count is 1 and `eventTime` is the correct choice), but I'd be ok with using the deprecated fields for this pr and raising an issue to add proper support later.

shopify-graveyard commented:
Yup, it took me a while to understand what was happening with k8s in `1.16.2` and `1.15.3` - I think the last commit covers the cases the case in `v1` where we don't have `lastTimeStamp` but have the `.series` field:

```ruby
[5] pry(Krane::KubernetesResource::Event)> FIELDS
=> [".involvedObject.kind",
 ".involvedObject.name",
 ".count",
 ".lastTimestamp",
 ".reason",
 ".message",
 ".eventTime",
 ".series"]
[6] pry(Krane::KubernetesResource::Event)> pieces
=> ["Pod",
 "pvc",
 "<no value>",
 "<no value>",
 "FailedScheduling",
 "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
 "2019-11-08T00:04:18.108148Z",
 "map[count:1 lastObservedTime:2019-11-08T00:04:21.109944Z]"]
```

douglas commented:
Yup, it took me a while to understand what was happening with k8s in `1.16.2` and `1.15.3` - I think the last commit covers the cases the case in `Event v1` where we don't have `lastTimeStamp` but have the `.series` field:

```ruby
[5] pry(Krane::KubernetesResource::Event)> FIELDS
=> [".involvedObject.kind",
 ".involvedObject.name",
 ".count",
 ".lastTimestamp",
 ".reason",
 ".message",
 ".eventTime",
 ".series"]
[6] pry(Krane::KubernetesResource::Event)> pieces
=> ["Pod",
 "pvc",
 "<no value>",
 "<no value>",
 "FailedScheduling",
 "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
 "2019-11-08T00:04:18.108148Z",
 "map[count:1 lastObservedTime:2019-11-08T00:04:21.109944Z]"]
```

dturn commented on lib/krane/kubernetes_resource.rb:
```diff
@@ -409,11 +411,26 @@ def self.go_template_for(kind, name)
+
+          # Find the right count and last_timestamp according to Events API version
+          count, last_timestamp = if series.present? && series != "<no value>" # Events/v1
+            count_regex = /count:(?<value>\S+?(?=\s))/
```
Any chance `JSON.parse(series)` works?

douglas replied:
hmmmm, I didn't knew about it and will try, thanks for the suggestion 👍

douglas replied:
Unfortunately, it does not work because `series` doesn't contain a valid json string, this is the content of the `series` field:

```ruby
[13] pry(Krane::KubernetesResource::Event)> blob
=> "PodENDFIELD--BEGINFIELDpvcENDFIELD--BEGINFIELD<no value>ENDFIELD--BEGINFIELD<no value>ENDFIELD--BEGINFIELDFailedSchedulingENDFIELD--BEGINFIELD0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.ENDFIELD--BEGINFIELD2019-11-08T18:56:08.886672ZENDFIELD--BEGINFIELD<no value>ENDEVENT--BEGINEVENTPodENDFIELD--BEGINFIELDpvcENDFIELD--BEGINFIELD<no value>ENDFIELD--BEGINFIELD<no value>ENDFIELD--BEGINFIELDFailedSchedulingENDFIELD--BEGINFIELD0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.ENDFIELD--BEGINFIELD2019-11-08T18:56:08.900067ZENDFIELD--BEGINFIELDmap[count:1 lastObservedTime:2019-11-08T18:56:10.183912Z]ENDEVENT--BEGINEVENT"
[14] pry(Krane::KubernetesResource::Event)> series
=> "map[count:1 lastObservedTime:2019-11-08T18:56:10.183912Z]"
```

dturn commented on lib/krane/kubernetes_resource.rb:
```diff
@@ -409,11 +411,26 @@ def self.go_template_for(kind, name)
+          elsif last_timestamp != "<no value>" # Events/v1beta1
+            [pieces[FIELDS.index(".count")], last_timestamp]
+          else # Fallback to eventTime
+            [1, pieces[FIELDS.index(".eventTime")]]
```
What does this case cover?

douglas replied:
I was trying to cover a case where `lastTimeStamp`, `deprecatedTimeStamp` and `series` where empty.

I will check what happens on `1.11` from `1.16` and will post here.

RyanBrushett commented on test/fixtures/invalid/bad_probe.yml:
```diff
@@ -1,9 +1,16 @@
-apiVersion: extensions/v1beta1 # keep as long as this group is valid
+apiVersion: apps/v1 # keep as long as this group is valid
```
Is this comment still needed?

douglas replied:
I don't think so - will remove it.

dturn commented on lib/krane/kubernetes_resource/daemon_set.rb:
```diff
@@ -78,9 +78,10 @@ def rollout_data
     def parent_of_pod?(pod_data)
       return false unless pod_data.dig("metadata", "ownerReferences")
+      template_generation = @instance_data["spec"]["templateGeneration"] ||
+        @instance_data["metadata"]["annotations"]["deprecated.daemonset.template.generation"]
```
Let's open an issue to find a long term solution here.

douglas replied:
Done: https://github.com/Shopify/kubernetes-deploy/issues/627

dturn replied:
Should we actually be using `dig` in case annotations don't exist?

douglas replied:
Good idea! Implemented 👍

dturn commented on lib/krane/kubernetes_resource.rb:
```diff
@@ -442,8 +473,6 @@ def global?
       @global || self.class::GLOBAL
     end
 
-    private
```
why was this deleted?

shopify-graveyard replied:
Sorry, I was trying something to replace `self.extract_count_timestamp` and forgot to put it back - fixed.

douglas replied:
Sorry, I was trying something to replace self.extract_count_timestamp and forgot to put it back - fixed.

dturn commented on lib/krane/kubernetes_resource.rb:
```diff
@@ -409,17 +413,44 @@ def self.go_template_for(kind, name)
+        deprecated_timestamp = pieces[FIELDS.index(".deprecatedLastTimestamp")]
+
+        # Find the right count and last_timestamp according to Kubernetes API and kubectl version
+        if deprecated_timestamp.present? && deprecated_timestamp != "<no value>"
```
Let's make `"<no value>"` be a constant

dturn commented on lib/krane/kubernetes_resource.rb:
```diff
@@ -409,17 +413,44 @@ def self.go_template_for(kind, name)
+        if deprecated_timestamp.present? && deprecated_timestamp != "<no value>"
+          # kubectl < 1.16 uses events.k8s.io/v1beta1, which has .deprecatedCount and
+          # .deprecatedLastTimestamp fields
+          [pieces[FIELDS.index(".deprecatedCount")], deprecated_timestamp]
```
I'm worried about coupling count to timestap. Given we've seen cases where they're both blank, it seems possible one could be but not the other one.  Can you also explain why you chose this ordering of cases?

douglas replied:
I broke the code to find the right event count and timestamp in two methods and I followed this order:

1) First deprecated fields `.deprecatedCount` and `.deprecatedLastTimestamp` because Kubernetes 1.16 is the only version that does not uses it, therefore, the deprecated fields are the ones that are more likely to appear.
2) If the above fields aren't present, I check for k8s 1.16 standard API fields
3) If the above fields aren't present, I check for the presence of .series field, which is present in k8s 1.1.6 as well
4) If none of those fields are present, I fallback to count as `1` and `.eventTime`, as in this case in Kubernetes `1.16`:

```
» kubectl get events -o json --namespace=k8sdeploy-test-pvc-no-pv-dafc5d11a073e407 --context=minikube --request-timeout=15
{
    "apiVersion": "v1",
    "items": [
        {
            "action": "Scheduling",
            "apiVersion": "v1",
            "eventTime": "2019-11-09T02:27:17.835909Z",
            "firstTimestamp": null,
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "Pod",
                "name": "pvc",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37187",
                "uid": "c07e276b-6ea9-4b27-801d-42949cfdc9c9"
            },
            "kind": "Event",
            "lastTimestamp": null,
            "message": "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:17Z",
                "name": "pvc.15d55d7d5c41277b",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37189",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/pvc.15d55d7d5c41277b",
                "uid": "1369fe84-67c1-4c43-826b-6ba2c33402ef"
            },
            "reason": "FailedScheduling",
            "reportingComponent": "default-scheduler",
            "reportingInstance": "default-scheduler-minikube",
            "source": {
                "component": "default-scheduler"
            },
            "type": "Warning"
        },
        {
            "action": "Scheduling",
            "apiVersion": "v1",
            "eventTime": "2019-11-09T02:27:17.851047Z",
            "firstTimestamp": null,
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "Pod",
                "name": "pvc",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37188",
                "uid": "c07e276b-6ea9-4b27-801d-42949cfdc9c9"
            },
            "kind": "Event",
            "lastTimestamp": null,
            "message": "0/1 nodes are available: 1 node(s) didn't find available persistent volumes to bind.",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:17Z",
                "name": "pvc.15d55d7d5d2828fa",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37194",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/pvc.15d55d7d5d2828fa",
                "uid": "8dc9a49e-66b2-444c-9c31-217cd8268ec7"
            },
            "reason": "FailedScheduling",
            "reportingComponent": "default-scheduler",
            "reportingInstance": "default-scheduler-minikube",
            "series": {
                "count": 1,
                "lastObservedTime": "2019-11-09T02:27:19.264307Z"
            },
            "source": {
                "component": "default-scheduler"
            },
            "type": "Warning"
        },
        {
            "apiVersion": "v1",
            "count": 3,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "with-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37173",
                "uid": "00c29fd4-8104-400f-b41f-d22012b01bee"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:44Z",
            "message": "waiting for first consumer to be created before binding",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "with-storage-class.15d55d7d0620c54f",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37238",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/with-storage-class.15d55d7d0620c54f",
                "uid": "a06ce861-756a-4660-9a7f-db53f083db15"
            },
            "reason": "WaitForFirstConsumer",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "persistentvolume-controller"
            },
            "type": "Normal"
        },
        {
            "apiVersion": "v1",
            "count": 2,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "without-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37176",
                "uid": "d0d2ff61-52e1-4850-a632-687b0f89ec8f"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:16Z",
            "message": "waiting for a volume to be created, either by external provisioner \"k8s.io/minikube-hostpath\" or manually created by system administrator",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "without-storage-class.15d55d7d07392317",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37179",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/without-storage-class.15d55d7d07392317",
                "uid": "b5d75be0-c579-407d-bd53-e9828ea7d4de"
            },
            "reason": "ExternalProvisioning",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "persistentvolume-controller"
            },
            "type": "Normal"
        },
        {
            "apiVersion": "v1",
            "count": 1,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "without-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37175",
                "uid": "d0d2ff61-52e1-4850-a632-687b0f89ec8f"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:16Z",
            "message": "External provisioner is provisioning volume for claim \"k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/without-storage-class\"",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "without-storage-class.15d55d7d07f1f2b0",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37180",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/without-storage-class.15d55d7d07f1f2b0",
                "uid": "6b5c8e3f-a7e2-4278-94dc-8afae68a94a6"
            },
            "reason": "Provisioning",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "k8s.io/minikube-hostpath c9d9d89f-01d8-11ea-9057-aa43d77eee0d"
            },
            "type": "Normal"
        },
        {
            "apiVersion": "v1",
            "count": 1,
            "eventTime": null,
            "firstTimestamp": "2019-11-09T02:27:16Z",
            "involvedObject": {
                "apiVersion": "v1",
                "kind": "PersistentVolumeClaim",
                "name": "without-storage-class",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37175",
                "uid": "d0d2ff61-52e1-4850-a632-687b0f89ec8f"
            },
            "kind": "Event",
            "lastTimestamp": "2019-11-09T02:27:16Z",
            "message": "Successfully provisioned volume pvc-d0d2ff61-52e1-4850-a632-687b0f89ec8f",
            "metadata": {
                "creationTimestamp": "2019-11-09T02:27:16Z",
                "name": "without-storage-class.15d55d7d08aef384",
                "namespace": "k8sdeploy-test-pvc-no-pv-dafc5d11a073e407",
                "resourceVersion": "37184",
                "selfLink": "/api/v1/namespaces/k8sdeploy-test-pvc-no-pv-dafc5d11a073e407/events/without-storage-class.15d55d7d08aef384",
                "uid": "ed4f557a-bd08-4726-8802-fdd173cc7a33"
            },
            "reason": "ProvisioningSucceeded",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "k8s.io/minikube-hostpath c9d9d89f-01d8-11ea-9057-aa43d77eee0d"
            },
            "type": "Normal"
        }
    ],
    "kind": "List",
    "metadata": {
        "resourceVersion": "",
        "selfLink": ""
    }
}
```

dturn replied:
I get what you're saying, but its confusing becasuse you don't seperate when you mean 
kubectl 1.16 preferred version vs resources version supported on k8s 1.16

douglas replied:
I reordered the extraction logic to follow what should correct k8s API order: currently supported fields, series for k8s 1.16, deprecated fields and fallback.

timothysmith0609 commented on lib/krane/kubeclient_builder.rb:
```diff
@@ -55,6 +55,14 @@ def build_policy_v1beta1_kubeclient(context)
+
     def build_apps_v1beta1_kubeclient(context)
       build_kubeclient(
         api_version: "v1beta1",
```
Do we need the apps_v1beta1 client anymore?

douglas replied:
Yeah, we still need it because of Ingress, which still uses `extensions/v1beta1`: 

https://github.com/Shopify/kubernetes-deploy/pull/598/files#diff-53358cd61c3052c22151ef2a24e68e9bL54 and https://github.com/Shopify/kubernetes-deploy/blob/fix_krane_k8s_api_1.16/test/fixtures/hello-cloud/web.yml.erb#L1.

timothysmith0609 replied:
That's `extensions_v1beta1_kubeclient`, though, not `apps_v1beta1_kubeclient`, unless I'm missing something?

douglas replied:
Oh, you're absolutely right! I don't think we need `apps_v1beta1_kubeclient` anymore, would it be a problem to remove it in a subsequent PR?

timothysmith0609 replied:
It's no problem, just trying to avoid cruft building up :)

douglas replied:
Just pushed a commit removing `apps_v1beta1_kubeclient` as it was a small change! Thanks for pointing it, Tim! ❤️

timothysmith0609 commented on lib/krane/kubernetes_resource.rb:
```diff
@@ -409,17 +414,61 @@ def self.go_template_for(kind, name)
+          # kubectl < 1.16 uses events.k8s.io/v1beta1, which has .deprecatedCount
+          deprecated_count
+        else
+          "1" # Fallback to 1 when all count fields are null
```
The API reference defines count as an integer, how come we're using a String?

jessie-JNing replied:
good point, that reference https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#event-v1-core does show it's an integer, 😕

douglas replied:
I'm using `"1"` to be consistent which is returned by `pieces[FIELDS.index(".count")]` or `pieces[FIELDS.index(".deprecatedCount")]`:

```ruby
kubectl get events --output=go-template={{range .items}}{{if and (eq .involvedObject.kind "Pod") (eq .involvedObject.name "pvc") (ne .reason "Started") (ne .reason "Created") (ne .reason "SuccessfulCreate") (ne .reason "Scheduled") (ne .reason "Pulling") (ne .reason "Pulled")}}{{.involvedObject.kind}}{{print "ENDFIELD--BEGINFIELD"}}{{.involvedObject.name}}{{print "ENDFIELD--BEGINFIELD"}}{{.count}}{{print "ENDFIELD--BEGINFIELD"}}{{.lastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.reason}}{{print "ENDFIELD--BEGINFIELD"}}{{.message}}{{print "ENDFIELD--BEGINFIELD"}}{{.eventTime}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedCount}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedLastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.series}}{{print "ENDEVENT--BEGINEVENT"}}{{end}}{{end}} --namespace=k8sdeploy-test-pvc-no-pv-4e96cca0a0d64155 --context=minikube --request-timeout=15

From: /Users/douglas/src/github.com/Shopify/kubernetes-deploy/lib/krane/kubernetes_resource.rb @ line 422 Krane::KubernetesResource::Event.extract_all_from_go_template_blob:

    414: def self.extract_all_from_go_template_blob(blob)
    415:   blob.split(EVENT_SEPARATOR).map do |event_blob|
    416:     pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
    417:     count = extract_event_count(pieces)
    418:     timestamp = extract_event_timestamp(pieces)
    419:
    420:     binding.pry
    421:
 => 422:     new(
    423:       subject_kind: pieces[FIELDS.index(".involvedObject.kind")],
    424:       subject_name: pieces[FIELDS.index(".involvedObject.name")],
    425:       count: count,
    426:       last_timestamp: timestamp,
    427:       reason: pieces[FIELDS.index(".reason")],
    428:       message: pieces[FIELDS.index(".message")]
    429:     )
    430:   end
    431: end
[1] pry(Krane::KubernetesResource::Event)> pieces[FIELDS.index(".count")]
=> "4"
```

After we get the correct `count` value, it is passed to `Event`initializer: 
 https://github.com/Shopify/kubernetes-deploy/blob/fix_krane_k8s_api_1.16/lib/krane/kubernetes_resource.rb#L478 which does `@count = count.to_i`, so we're still respecting k8s API.

douglas replied:
I'm using `"1"` to be consistent which is returned by `pieces[FIELDS.index(".count")]` or `pieces[FIELDS.index(".deprecatedCount")]`:

```ruby
kubectl get events --output=go-template={{range .items}}{{if and (eq .involvedObject.kind "Pod") (eq .involvedObject.name "pvc") (ne .reason "Started") (ne .reason "Created") (ne .reason "SuccessfulCreate") (ne .reason "Scheduled") (ne .reason "Pulling") (ne .reason "Pulled")}}{{.involvedObject.kind}}{{print "ENDFIELD--BEGINFIELD"}}{{.involvedObject.name}}{{print "ENDFIELD--BEGINFIELD"}}{{.count}}{{print "ENDFIELD--BEGINFIELD"}}{{.lastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.reason}}{{print "ENDFIELD--BEGINFIELD"}}{{.message}}{{print "ENDFIELD--BEGINFIELD"}}{{.eventTime}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedCount}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedLastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.series}}{{print "ENDEVENT--BEGINEVENT"}}{{end}}{{end}} --namespace=k8sdeploy-test-pvc-no-pv-4e96cca0a0d64155 --context=minikube --request-timeout=15

From: /Users/douglas/src/github.com/Shopify/kubernetes-deploy/lib/krane/kubernetes_resource.rb @ line 422 Krane::KubernetesResource::Event.extract_all_from_go_template_blob:

    414: def self.extract_all_from_go_template_blob(blob)
    415:   blob.split(EVENT_SEPARATOR).map do |event_blob|
    416:     pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
    417:     count = extract_event_count(pieces)
    418:     timestamp = extract_event_timestamp(pieces)
    419:
    420:     binding.pry
    421:
 => 422:     new(
    423:       subject_kind: pieces[FIELDS.index(".involvedObject.kind")],
    424:       subject_name: pieces[FIELDS.index(".involvedObject.name")],
    425:       count: count,
    426:       last_timestamp: timestamp,
    427:       reason: pieces[FIELDS.index(".reason")],
    428:       message: pieces[FIELDS.index(".message")]
    429:     )
    430:   end
    431: end
[1] pry(Krane::KubernetesResource::Event)> pieces[FIELDS.index(".count")]
=> "4"
```

After we get the correct `count` value, it is passed to `Event`initializer: 
 https://github.com/Shopify/kubernetes-deploy/blob/fix_krane_k8s_api_1.16/lib/krane/kubernetes_resource.rb#L478 which does `@count = count.to_i`, so we're still respecting k8s API.

jessie-JNing replied:
Thanks for the prove, good explanation @douglas

timothysmith0609 commented on test/helpers/fixture_set.rb:
```diff
@@ -28,7 +28,14 @@ def app_name
+        apps_v1_kubeclient
+      else
+        kubeclient
+      end
```
Do we need to worry about the v1beta1 client? From the docs, it appears that apps/v1 should work in this case, no? Might be worth trying out if you haven't already

douglas replied:
Yup, we need to care because `Ingress` kind still uses it 😢

jessie-JNing commented on test/fixtures/ejson-cloud/web.yaml:
```diff
@@ -1,10 +1,17 @@
 kind: Deployment
 metadata:
   name: web
+  labels:
```
Is it because `labels` and `selector` become mandatory fields in `apps/v1`?

douglas replied:
Yeah, this is why I had to add it to a bunch of templates.

jessie-JNing commented on README.md:
````diff
@@ -524,7 +524,7 @@ The following command will restart all pods in the `web` and `jobs` deployments:
 
 ```yaml
-apiVersion: apps/v1beta1
+apiVersion: apps/v1
````
If the users are under `v1.16`, apiVersion should still be `apps/v1beta1`, or it's compatible? I'm a little confused

douglas replied:
This is one of the core changes for `k8s` 1.16, for `deployments, statefulsets, daemonsets and replicasets` require `apps/v1` and will fail when users try to use `apps/v1beta1` with `1.16`.

That is why I had to change everything from `apps/v1beta1` to `apps/v1`, is it clear now?

jessie-JNing replied:
From then readme point of view, if the users are in k8s `v1.15`,  will `apps/v1` work for them?

douglas replied:
Yes, `apps/v1` works (for some since `k8s` 1.9, this is why our tests pass in buildkite for the 1.11 and above for this branch.

Here is the documentation that mentions it: https://kubernetes.io/blog/2019/09/18/kubernetes-1-16-release-announcement/
